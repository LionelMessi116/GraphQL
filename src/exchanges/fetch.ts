/* eslint-disable @typescript-eslint/no-use-before-define */
import { print } from 'graphql';
import { filter, make, merge, mergeMap, pipe, share, takeUntil } from 'wonka';
import { Exchange, Operation, OperationResult } from '../types';
import { addMetadata, makeResult, makeErrorResult } from '../utils';

/** A default exchange for fetching GraphQL requests. */
export const fetchExchange: Exchange = ({ forward }) => {
  const isOperationFetchable = (operation: Operation) => {
    const { operationName } = operation;
    return operationName === 'query' || operationName === 'mutation';
  };

  return ops$ => {
    const sharedOps$ = share(ops$);
    const fetchResults$ = pipe(
      sharedOps$,
      filter(isOperationFetchable),
      mergeMap(operation => {
        const { key } = operation;
        const teardown$ = pipe(
          sharedOps$,
          filter(op => op.operationName === 'teardown' && op.key === key)
        );

        return pipe(
          createFetchSource(operation),
          takeUntil(teardown$)
        );
      })
    );

    const forward$ = pipe(
      sharedOps$,
      filter(op => !isOperationFetchable(op)),
      forward
    );

    return merge([fetchResults$, forward$]);
  };
};

const createFetchSource = (operation: Operation) => {
  if (
    process.env.NODE_ENV !== 'production' &&
    operation.operationName === 'subscription'
  ) {
    throw new Error(
      'Received a subscription operation in the httpExchange. You are probably trying to create a subscription. Have you added a subscriptionExchange?'
    );
  }

  return make<OperationResult>(([next, complete]) => {
    const abortController =
      typeof AbortController !== 'undefined'
        ? new AbortController()
        : undefined;

    const { context } = operation;

    const extraOptions =
      typeof context.fetchOptions === 'function'
        ? context.fetchOptions()
        : context.fetchOptions || {};

    const fetchOptions = {
      body: JSON.stringify({
        query: print(operation.query),
        variables: operation.variables,
      }),
      method: 'POST',
      ...extraOptions,
      headers: {
        'content-type': 'application/json',
        ...extraOptions.headers,
      },
      signal:
        abortController !== undefined ? abortController.signal : undefined,
    };

    const startTime = Date.now();

    executeFetch(operation, fetchOptions).then(result => {
      if (result !== undefined) {
        if (process.env.NODE_ENV !== 'production') {
          next({
            ...result,
            operation: addMetadata(result.operation, {
              networkLatency: Date.now() - startTime,
            }),
          });
        } else {
          next(result);
        }
      }

      complete();
    });

    return () => {
      if (abortController !== undefined) {
        abortController.abort();
      }
    };
  });
};

const executeFetch = (operation: Operation, opts: RequestInit) => {
  const { url, fetch: fetcher } = operation.context;

  let response: Response | undefined;

  return (fetcher || fetch)(url, opts)
    .then(res => {
      const { status } = res;
      const statusRangeEnd = opts.redirect === 'manual' ? 400 : 300;
      response = res;

      if (status < 200 || status >= statusRangeEnd) {
        throw new Error(res.statusText);
      } else {
        return res.json();
      }
    })
    .then(result => makeResult(operation, result, response))
    .catch(err => {
      if (err.name !== 'AbortError') {
        return makeErrorResult(operation, err, response);
      }
    });
};
