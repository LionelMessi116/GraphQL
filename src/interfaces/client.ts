import Observable from 'zen-observable-ts';

import { ICache } from './cache';
import { IExchangeResult } from './exchange';
import { IQuery } from './query';

export interface IClient {
  cache: ICache;
  executeQuery$(
    queryObject: IQuery,
    skipCache: boolean
  ): Observable<IExchangeResult>;
  executeQuery(
    queryObject: IQuery,
    skipCache: boolean
  ): Promise<IExchangeResult>;
  executeMutation$(mutationObject: IQuery): Observable<IExchangeResult['data']>;
  executeMutation(mutationObject: IQuery): Promise<IExchangeResult['data']>;
  refreshAllFromCache(): void;
  subscribe(
    callback: (changedTypes: string[], reponse: object) => void
  ): () => void;
}
