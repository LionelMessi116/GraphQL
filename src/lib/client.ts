import {
  filter,
  makeSubject,
  pipe,
  publish,
  share,
  Source,
  Subject,
  subscribe,
  take,
} from 'wonka';

import * as uuid from 'uuid';
import {
  cacheExchange,
  dedupeExchange,
  fetchExchange,
  subscriptionExchange,
} from '../exchanges';
import { CombinedError, hashString } from '../lib';
import {
  Client,
  ClientInstance,
  ClientOptions,
  CreateClientInstanceOpts,
  Exchange,
  ExchangeResult,
  Mutation,
  Operation,
  OperationType,
  Query,
  Subscription,
  SubscriptionStreamUpdate,
} from '../types';

type SubscriptionMap = Map<string, () => void>;

const defaultExchanges = [dedupeExchange, cacheExchange, fetchExchange];

/** Function for creating an application wide [Client]{@link Client}. */
export const createClient = (opts: ClientOptions): Client => {
  /** Cache of all instance subscriptions */
  const subscriptions = new Map<string, SubscriptionMap>();
  const querySubscriptions = new Map<string, SubscriptionMap>();

  /** Main subject for emitting operations */
  const subject = makeSubject<Operation>();

  const exchanges =
    opts.exchanges !== undefined ? opts.exchanges : defaultExchanges;

  /** Main subject stream for listening to operation responses */
  const subject$ = pipe(
    subject[0],
    pipeExchanges(exchanges, subject),
    share
  );

  publish(subject$);

  const fetchOptions =
    typeof opts.fetchOptions === 'function'
      ? opts.fetchOptions()
      : opts.fetchOptions;

  /** Convert a query to an operation type */
  const createOperation = (
    type: OperationType,
    query: Query,
    options: object = {}
  ) => ({
    ...query,
    key: hashString(JSON.stringify(query)),
    operationName: type,
    context: {
      url: opts.url,
      fetchOptions,
      forwardSubscription: opts.forwardSubscription,
      ...options,
    },
  });

  const executeOperation = (operation: Operation) => subject[1](operation);

  /** Function to manage subscriptions on a Consumer by Consumer basis */
  const createInstance = (
    instanceOpts: CreateClientInstanceOpts
  ): ClientInstance => {
    const id = uuid.v4();

    const getSubscriptions = () =>
      subscriptions.has(id) ? subscriptions.get(id) : (new Map() as SubscriptionMap);

    const updateSubscriptions = (subs: SubscriptionMap) =>
      subscriptions.set(id, subs);

    const executeQuery = (query: Query, force?: boolean) => {
      const operation = createOperation(OperationType.Query, query, { force });

      const outboundKey = `${operation.key}-out`;
      const inboundKey = `${operation.key}-in`;

      const instanceSubscriptions = getSubscriptions();
      if (!instanceSubscriptions.has(outboundKey)) {
        /** Notify component when fetching query is occurring */
        const [outboundSub] = pipe(
          subject[0],
          filter(op => op.key === operation.key),
          subscribe(() => instanceOpts.onChange({ fetching: true }))
        );

        instanceSubscriptions.set(outboundKey, outboundSub);
      }

      if (!instanceSubscriptions.has(inboundKey)) {
        /** Notify component when query operation has returned */
        const [inboundSub] = pipe(
          subject$,
          filter(result => result.operation.key === operation.key),
          subscribe(response => {
            instanceOpts.onChange({
              data: response.data,
              error: response.error,
              fetching: false,
            });
          })
        );

        instanceSubscriptions.set(inboundKey, inboundSub);
      }

      updateSubscriptions(instanceSubscriptions);
      executeOperation(operation);
    };

    const executeMutation = (query: Mutation) => {
      const operation = createOperation(OperationType.Mutation, query);
      executeOperation(operation);
    };

    const executeSubscription = (query: Subscription) => {
      const operation = createOperation(OperationType.Subscription, query);
      const subscriptionSubject = makeSubject<Operation>();

      querySubscriptions.set(query.query, subscriptionSubject);

      /** Notify component when subscription operation has been given a value */
      subscriptionSubject
        .pipe(
          pipeExchanges(
            opts.exchanges !== undefined ? opts.exchanges : defaultExchanges,
            subject
          )
        )
        .subscribe((response: SubscriptionStreamUpdate) => {
          instanceOpts.onSubscriptionUpdate({
            data: response.data.data,
            error: Array.isArray(response.data.errors)
              ? new CombinedError({
                  graphQLErrors: response.data.errors,
                  response,
                })
              : response.error,
          });
        });

      executeOperation(operation);
    };

    const executeUnsubscribeSubscription = (query: Subscription) => {
      const operation = createOperation(OperationType.Subscription, query, {
        unsubscribe: true,
      });

      // remove cached rx subscription
      const activeSubscriptionSubject = querySubscriptions.get(query.query);

      if (activeSubscriptionSubject !== undefined) {
        querySubscriptions.delete(query.query);
        executeOperation(operation);

        // This removes the previous existence of a subscription from being part of the subject
        // without this, any unsubscribe -> subscribe will have an n^2 effect.
        activeSubscriptionSubject.unsubscribe();
      }
    };

    const unsubscribe = () => {
      const subs = getRxSubscriptions();
      if (subs === undefined) {
        return;
      }

      [...subs.values()].forEach(unsub => unsub());
    };

    return {
      executeQuery,
      executeMutation,
      executeSubscription,
      executeUnsubscribeSubscription,
      unsubscribe,
    };
  };

  return {
    createInstance,
  };
};

/** Create pipe of exchanges */
const pipeExchanges = (exchanges: Exchange[], subject: Subject<Operation>) => (
  operation$: Source<Operation>
): Source<ExchangeResult> => {
  /** Recursively pipe to each exchange */
  const callExchanges = (value: Source<Operation>, index: number = 0) => {
    if (exchanges.length < index) {
      return value;
    }

    const currentExchange = exchanges[index];
    return currentExchange({
      forward: val => callExchanges(val, index + 1),
      subject,
    })(value);
  };

  return callExchanges(operation$, 0);
};
