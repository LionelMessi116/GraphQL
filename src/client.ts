import { filter, makeSubject, onStart, pipe, share, Source, take } from 'wonka';
import { composeExchanges, defaultExchanges } from './exchanges';
import { hashString } from './lib';

import {
  Exchange,
  GraphqlMutation,
  GraphqlQuery,
  GraphQLRequest,
  GraphqlSubscription,
  Operation,
  OperationContext,
  OperationResult,
  OperationType,
} from './types';

/** Options for configuring the URQL [client]{@link Client}. */
export interface ClientOptions {
  /** Target endpoint URL such as `https://my-target:8080/graphql`. */
  url: string;
  /** Any additional options to pass to fetch. */
  fetchOptions?: RequestInit | (() => RequestInit);
  /** An ordered array of Exchanges. */
  exchanges?: Exchange[];
}

interface ActiveResultSources {
  [operationKey: string]: Source<OperationResult>;
}

export const createClient = (opts: ClientOptions) => new Client(opts);

/** The URQL application-wide client library. Each execute method starts a GraphQL request and returns a stream of results. */
export class Client implements ClientOptions {
  // These are variables derived from ClientOptions
  url: string;
  fetchOptions: RequestInit;
  exchanges: Exchange[];

  // These are internals to be used to keep track of operations
  dispatchOperation: (operation: Operation) => void;
  operations$: Source<Operation>;
  results$: Source<OperationResult>;
  activeResultSources = Object.create(null) as ActiveResultSources;

  constructor(opts: ClientOptions) {
    this.url = opts.url;

    this.fetchOptions =
      typeof opts.fetchOptions === 'function'
        ? opts.fetchOptions()
        : opts.fetchOptions || {};

    this.exchanges =
      opts.exchanges !== undefined ? opts.exchanges : defaultExchanges;

    // This subject forms the input of operations; executeOperation may be
    // called to dispatch a new operation on the subject
    const [operations$, nextOperation] = makeSubject<Operation>();
    this.operations$ = operations$;
    this.dispatchOperation = nextOperation;

    // All operations run through the exchanges in a pipeline-like fashion
    // and this observable then combines all their results
    this.results$ = share(
      composeExchanges(this, this.exchanges)(this.operations$)
    );
  }

  private createOperationContext = (
    opts?: Partial<OperationContext>
  ): OperationContext => ({
    url: this.url,
    fetchOptions: this.fetchOptions,
    ...opts,
  });

  private createRequestOperation = (
    type: OperationType,
    query: GraphQLRequest,
    opts?: Partial<OperationContext>
  ): Operation => ({
    ...query,
    key: hashString(JSON.stringify(query)),
    operationName: type,
    context: this.createOperationContext(opts),
  });

  /** Deletes an active operation's result observable and sends a teardown signal through the exchange pipeline */
  // private teardownOperation(operation: Operation) {
  //   delete this.activeResultSources[operation.key];
  //   this.dispatchOperation({ ...operation, operationName: 'teardown' });
  // }

  /** Executes an Operation by sending it through the exchange pipeline It returns an observable that emits all related exchange results and keeps track of this observable's subscribers. A teardown signal will be emitted when no subscribers are listening anymore. */
  executeRequestOperation(operation: Operation): Source<OperationResult> {
    const { key, operationName } = operation;

    const operationResults$ = pipe(
      this.results$,
      filter(res => res.operation.key === key)
    );

    if (operationName === 'mutation') {
      // A mutation is always limited to just a single result and is never shared
      return pipe(
        operationResults$,
        onStart<OperationResult>(() => this.dispatchOperation(operation)),
        take(1)
      );
    }

    if (key in this.activeResultSources) {
      // Reuse active result observable when it's available
      return this.activeResultSources[key];
    }

    return (this.activeResultSources[key] = pipe(
      operationResults$,
      onStart<OperationResult>(() => this.dispatchOperation(operation)),
      // onEnd<OperationResult>(() => this.teardownOperation(operation)),
      share
    ));
  }

  reexecuteOperation = (operation: Operation) => {
    // Reexecute operation only if any subscribers are still subscribed to the
    // operation's exchange results
    if (this.activeResultSources[operation.key] !== undefined) {
      this.dispatchOperation(operation);
    }
  };

  executeQuery = (
    query: GraphqlQuery,
    opts?: Partial<OperationContext>
  ): Source<OperationResult> => {
    const operation = this.createRequestOperation('query', query, opts);
    return this.executeRequestOperation(operation);
  };

  executeSubscription = (
    query: GraphqlSubscription,
    opts?: Partial<OperationContext>
  ): Source<OperationResult> => {
    const operation = this.createRequestOperation('subscription', query, opts);
    return this.executeRequestOperation(operation);
  };

  executeMutation = (
    query: GraphqlMutation,
    opts?: Partial<OperationContext>
  ): Source<OperationResult> => {
    const operation = this.createRequestOperation('mutation', query, opts);
    return this.executeRequestOperation(operation);
  };
}
