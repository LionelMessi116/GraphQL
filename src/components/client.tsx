import { Component, ReactNode } from 'react';
import { CombinedError } from '../lib';
import {
  ChildArgs,
  Client,
  ClientInstance,
  Mutation,
  Query,
  StreamUpdate,
} from '../types';

export interface ClientState<MutationDeclarations> {
  client: ClientInstance;
  fetching: boolean;
  error?: Error | CombinedError | CombinedError[];
  data?: any | any[];
  mutations?: { [type in keyof MutationDeclarations]: (vals: any) => void };
}

export interface ClientProps<MutationDeclarations> {
  client: Client;
  children: (obj: ChildArgs<MutationDeclarations>) => ReactNode;
  query?: Query;
  mutations?: { [type in keyof MutationDeclarations]: Mutation };
}

/** Component responsible for implementing the [Client]{@link Client} utility in React. */
export class UrqlClient<MutationDeclarations> extends Component<
  ClientProps<MutationDeclarations>,
  ClientState<MutationDeclarations>
> {
  constructor(props: ClientProps<MutationDeclarations>) {
    super(props);

    this.state = {
      client: props.client.createInstance({
        onChange: result => this.onStreamUpdate(result),
      }),
      data: undefined,
      error: undefined,
      fetching: true,
      mutations: this.getMutatorFunctions(),
    };
  }

  public componentDidMount() {
    if (this.props.query !== undefined) {
      this.state.client.executeQuery(this.props.query);
    }
  }

  public componentDidUpdate(prevProps: ClientProps<MutationDeclarations>) {
    if (
      JSON.stringify(prevProps.mutations) !==
      JSON.stringify(this.props.mutations)
    ) {
      this.setState({ mutations: this.getMutatorFunctions() });
    }

    if (
      this.props.query !== undefined &&
      JSON.stringify(prevProps.query) !== JSON.stringify(this.props.query)
    ) {
      this.state.client.executeQuery(this.props.query);
    }
  }

  public componentWillUnmount() {
    this.state.client.unsubscribe();
  }

  public render() {
    if (typeof this.props.children !== 'function') {
      return null;
    }

    return this.props.children({
      fetching: this.state.fetching,
      error: this.state.error,
      data: this.state.data,
      mutations: this.state.mutations,
      refetch: (noCache: boolean = false) =>
        this.props.query !== undefined
          ? this.state.client.executeQuery(this.props.query, noCache)
          : undefined,
    });
  }

  private getMutatorFunctions() {
    if (this.props.mutations === undefined) {
      return undefined;
    }

    const mutations = Object.keys(this.props.mutations).reduce(
      (prev, key) => ({
        ...prev,
        [key]: (variables: any) =>
          this.state.client.executeMutation({
            // @ts-ignore -  mutation key definitely exists
            ...this.props.mutations[key],
            variables,
          }),
      }),
      {}
    ) as ClientState<MutationDeclarations>['mutations'];

    return mutations;
  }

  private onStreamUpdate(updated: StreamUpdate) {
    this.setState({
      fetching: updated.fetching,
      error: updated.error,
      data: updated.data,
    });
  }
}
