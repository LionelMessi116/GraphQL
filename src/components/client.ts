import { Component, ReactNode } from 'react';
import { empty, filter, makeSubject, map, pipe, Source, subscribe, switchAll, toPromise } from 'wonka';

import { Client } from '../lib/client';
import { CombinedError } from '../lib/error';
import { ConnectProps } from './connect';

import {
  ChildArgs,
  ExchangeResult,
  Mutation,
  Query,
  Subscription
} from '../types';

export type Mutations<MutationDeclarations> = {
  [type in keyof MutationDeclarations]: Mutation
};

export type MutationFns<MutationDeclarations> = {
  [type in keyof MutationDeclarations]: (variables?: object) => Promise<ExchangeResult>
};

export interface ClientState<MutationDeclarations> {
  loaded: boolean;
  error?: CombinedError;
  data?: any;
  mutations?: MutationFns<MutationDeclarations>;
}

export interface ClientProps<MutationDeclarations> extends ConnectProps<MutationDeclarations> {
  client: Client;
}

/** Component responsible for implementing the [Client]{@link Client} utility in React. */
export class UrqlClient<MutationDeclarations> extends Component<
  ClientProps<MutationDeclarations>,
  ClientState<MutationDeclarations>
> {
  unsubscribe: () => void;
  nextQueryProp: (query?: Query) => void;
  nextSubscriptionProp: (subscription?: Subscription) => void;
  mutations = getMutationFns(this.props.client, this.props.mutations);

  state = {
    data: undefined,
    error: undefined,
    loaded: false,
  };

  constructor(props) {
    super(props);

    if (process.env.NODE_ENV !== 'production' && props.subscription && !props.updateSubscription) {
      throw new Error(
        'You instantiated an Urql Client component with a subscription but forgot an updateSubscription prop. updateSubscription callbacks are required to work with subscriptions.'
      );
    }

    const [queryProp$, nextQueryProp] = makeSubject<void | Query>();
    const [subscriptionProp$, nextSubscriptionProp] = makeSubject<void | Subscription>();
    this.nextQueryProp = nextQueryProp;
    this.nextSubscriptionProp = nextSubscriptionProp;

    const queryResults$ = makeQueryResults$(props.client, queryProp$);
    const subscriptionResults$ = makeSubscriptionResults$(props.client, subscriptionProp$);

    const [queryTeardown] = pipe(queryResults$, subscribe(this.onQueryUpdate));
    const [subscriptionTeardown] = pipe(queryResults$, subscribe(this.onSubscriptionUpdate));

    this.unsubscribe = () => {
      queryTeardown();
      subscriptionTeardown();
    };
  }

  private onQueryUpdate(result: ExchangeResult) {
    this.setState({
      loaded: true,
      error: result.error,
      data: result.data
    });
  }

  private onSubscriptionUpdate({ data, error }: ExchangeResult) {
    if (process.env.NODE_ENV !== 'production' && error) {
      throw error;
    }

    this.setState({
      loaded: true,
      error,
      data: this.props.updateSubscription
        ? this.props.updateSubscription(this.state.data, data)
        : this.state.data,
    });
  }

  componentDidMount() {
    this.nextQueryProp(this.props.query);
    this.nextSubscriptionProp(this.props.subscription);
  }

  componentDidUpdate(prevProps: ClientProps<MutationDeclarations>) {
    this.nextQueryProp(this.props.query);
    this.nextSubscriptionProp(this.props.subscription);

    if (prevProps.mutations !== this.props.mutations) {
      this.mutations = getMutationFns(this.props.client, this.props.mutations);
    }
  }

  componentWillUnmount() {
    this.unsubscribe();
  }

  render() {
    if (typeof this.props.children !== 'function') {
      return null;
    }

    return this.props.children({
      loaded: this.state.loaded,
      error: this.state.error,
      data: this.state.data,
      mutations: this.mutations || undefined
    });
  }
}

const makeQueryResults$ = (
  client: Client,
  queryProp$: Source<void | Query>
): Source<ExchangeResult> => {
  const noopQuery = (empty as Source<ExchangeResult>);
  let lastResults$: void | Source<ExchangeResult>;

  return pipe(
    queryProp$,
    map(query => {
      return query === undefined
        ? noopQuery
        : client.executeQuery(query);
    }),
    filter(x => {
      const isDistinct = x !== lastResults$;
      if (isDistinct) { lastResults$ = x; }
      return isDistinct;
    }),
    switchAll
  );
};

const makeSubscriptionResults$ = (
  client: Client,
  subscriptionProp$: Source<void | Subscription>
): Source<ExchangeResult> => {
  const noopSubscription = (empty as Source<ExchangeResult>);
  let lastResults$: void | Source<ExchangeResult>;

  return pipe(
    subscriptionProp$,
    map(query => {
      return query === undefined
        ? noopSubscription
        : client.executeSubscription(query);
    }),
    filter(x => {
      const isDistinct = x !== lastResults$;
      if (isDistinct) { lastResults$ = x; }
      return isDistinct;
    }),
    switchAll
  );
};

const getMutationFns = <T>(
  client: Client,
  mutations?: Mutations<T>
): void | MutationFns<T> => {
  if (mutations === undefined) {
    return undefined;
  }

  return Object.keys(mutations).reduce(
    (acc, key) => {
      acc[key] = (variables?: object) => pipe(
        client.executeMutation({
          ...mutations[key],
          variables,
        }),
        toPromise
      );

      return acc;
    },
    {} as MutationFns<T>
  );
}
