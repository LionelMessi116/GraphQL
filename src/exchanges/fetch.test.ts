import { empty, fromValue, pipe, Source, subscribe, toPromise } from 'wonka';
import { Client } from '../client';
import { queryOperation } from '../test-utils';
import { OperationResult, OperationType } from '../types';
import { fetchExchange } from './fetch';

const fetch = (global as any).fetch as jest.Mock;
const abort = jest.fn();

const abortError = new Error();
abortError.name = 'AbortError';

beforeAll(() => {
  (global as any).AbortController = function AbortController() {
    this.signal = undefined;
    this.abort = abort;
  };
});

beforeEach(() => {
  fetch.mockClear();
  abort.mockClear();
});

afterAll(() => {
  (global as any).AbortController = undefined;
});

const response = {
  status: 200,
  data: {
    data: {
      user: 1200,
    },
  },
};

const exchangeArgs = {
  forward: () => empty as Source<OperationResult>,
  client: {} as Client,
};

it('returns response data from fetch', async () => {
  fetch.mockResolvedValue({
    status: 200,
    json: jest.fn().mockResolvedValue(response),
  });

  const fetchOptions = jest.fn().mockReturnValue({});

  const data = await pipe(
    fromValue({
      ...queryOperation,
      context: {
        ...queryOperation.context,
        fetchOptions,
      },
    }),
    fetchExchange(exchangeArgs),
    toPromise
  );

  expect(data).toMatchSnapshot();
  expect(fetchOptions).toHaveBeenCalled();
});

it('returns error data from fetch', async () => {
  fetch.mockResolvedValue({
    status: 400,
    json: jest.fn().mockResolvedValue(response),
  });

  const data = await pipe(
    fromValue(queryOperation),
    fetchExchange(exchangeArgs),
    toPromise
  );

  expect(data).toMatchSnapshot();
});

it('calls cancel when the Observable is cancelled', () => {
  fetch.mockReturnValue(Promise.reject(abortError));

  const [unsubscribe] = pipe(
    fromValue(queryOperation),
    fetchExchange(exchangeArgs),
    subscribe(fail)
  );

  unsubscribe();
  expect(fetch).toHaveBeenCalledTimes(1);
  expect(abort).toHaveBeenCalledTimes(1);
});

it('should not start the operation when teardown gets passed', () => {
  pipe(
    fromValue({
      ...queryOperation,
      operationName: 'teardown' as OperationType,
    }),
    fetchExchange(exchangeArgs),
    subscribe(fail)
  );

  expect(fetch).toHaveBeenCalledTimes(0);
  expect(abort).toHaveBeenCalledTimes(0);
});
