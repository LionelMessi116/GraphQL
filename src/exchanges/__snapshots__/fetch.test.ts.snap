// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on error returns error data 1`] = `
Object {
  "data": undefined,
  "error": [CombinedError: [Network] ],
  "operation": Object {
    "context": Object {
      "fetchOptions": Object {
        "method": "POST",
      },
      "meta": Object {
        "networkLatency": 100,
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 2,
    "operationName": "query",
    "query": Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "getUser",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "name",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "name",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "firstName",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "lastName",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "name",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 124,
        "start": 0,
      },
    },
    "variables": Object {
      "name": "Clara",
    },
  },
}
`;

exports[`on error returns error data with status 400 and manual redirect mode 1`] = `
Object {
  "data": undefined,
  "error": [CombinedError: [Network] ],
  "operation": Object {
    "context": Object {
      "fetchOptions": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Object {
              "redirect": "manual",
            },
          },
        ],
      },
      "meta": Object {
        "networkLatency": 100,
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 2,
    "operationName": "query",
    "query": Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "getUser",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "name",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "name",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "firstName",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "lastName",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "name",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 124,
        "start": 0,
      },
    },
    "variables": Object {
      "name": "Clara",
    },
  },
}
`;

exports[`on success returns response data 1`] = `
Object {
  "data": Object {
    "data": Object {
      "user": 1200,
    },
  },
  "error": undefined,
  "operation": Object {
    "context": Object {
      "fetchOptions": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Object {},
          },
        ],
      },
      "meta": Object {
        "networkLatency": 100,
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 2,
    "operationName": "query",
    "query": Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "getUser",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "name",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "name",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "firstName",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "lastName",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "name",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 124,
        "start": 0,
      },
    },
    "variables": Object {
      "name": "Clara",
    },
  },
}
`;
